File: test/clients/go/partitioning_test.go
// Partitioning + clustering coverage:
//   - Ingestion-time partitioning + hidden _PARTITIONTIME column
//   - Column partitioning (TimePartitioning.Field)
//   - Clustering metadata
//   - is_partitioning_column via INFORMATION_SCHEMA.COLUMNS
package bqlocal_test

import (
	"context"
	"testing"

	"cloud.google.com/go/bigquery"
)

func TestIngestionTimePartitioning(t *testing.T) {
	ctx := context.Background()
	client, _ := newClient(t)
	if err := client.Dataset("ds").Create(ctx, &bigquery.DatasetMetadata{}); err != nil {
		t.Fatalf("dataset: %v", err)
	}
	table := client.Dataset("ds").Table("events")
	if err := table.Create(ctx, &bigquery.TableMetadata{
		Schema:           bigquery.Schema{{Name: "kind", Type: bigquery.StringFieldType}},
		TimePartitioning: &bigquery.TimePartitioning{Type: bigquery.DayPartitioningType},
	}); err != nil {
		t.Fatalf("create table: %v", err)
	}
	if err := table.Inserter().Put(ctx, []*struct {
		Kind string `bigquery:"kind"`
	}{{Kind: "click"}, {Kind: "view"}}); err != nil {
		t.Fatalf("insert: %v", err)
	}
	rows, err := readAll(t, client.Query(
		"SELECT kind, _PARTITIONTIME IS NOT NULL AS has_ts "+
			"FROM `ds.events` ORDER BY kind"))
	if err != nil {
		t.Fatalf("query: %v", err)
	}
	for _, r := range rows {
		if !r[1].(bool) {
			t.Errorf("_PARTITIONTIME missing for %v", r[0])
		}
	}
}

func TestPartitionDateFilterToday(t *testing.T) {
	ctx := context.Background()
	client, _ := newClient(t)
	if err := client.Dataset("ds").Create(ctx, &bigquery.DatasetMetadata{}); err != nil {
		t.Fatalf("dataset: %v", err)
	}
	table := client.Dataset("ds").Table("events")
	if err := table.Create(ctx, &bigquery.TableMetadata{
		Schema:           bigquery.Schema{{Name: "kind", Type: bigquery.StringFieldType}},
		TimePartitioning: &bigquery.TimePartitioning{Type: bigquery.DayPartitioningType},
	}); err != nil {
		t.Fatalf("create: %v", err)
	}
	if err := table.Inserter().Put(ctx, []*struct {
		Kind string `bigquery:"kind"`
	}{{Kind: "a"}, {Kind: "b"}}); err != nil {
		t.Fatalf("insert: %v", err)
	}
	rows, err := readAll(t, client.Query(
		"SELECT count(*)::INT64 AS n FROM `ds.events` "+
			"WHERE _PARTITIONDATE = CURRENT_DATE()"))
	if err != nil {
		t.Fatalf("query: %v", err)
	}
	if rows[0][0].(int64) != 2 {
		t.Errorf("partition count = %v, want 2", rows[0][0])
	}
}

func TestColumnPartitioningRoundTrip(t *testing.T) {
	ctx := context.Background()
	client, _ := newClient(t)
	if err := client.Dataset("ds").Create(ctx, &bigquery.DatasetMetadata{}); err != nil {
		t.Fatalf("dataset: %v", err)
	}
	table := client.Dataset("ds").Table("orders")
	if err := table.Create(ctx, &bigquery.TableMetadata{
		Schema: bigquery.Schema{
			{Name: "id", Type: bigquery.IntegerFieldType},
			{Name: "order_date", Type: bigquery.DateFieldType},
			{Name: "amount", Type: bigquery.FloatFieldType},
		},
		TimePartitioning: &bigquery.TimePartitioning{
			Type:  bigquery.DayPartitioningType,
			Field: "order_date",
		},
	}); err != nil {
		t.Fatalf("create: %v", err)
	}
	md, err := table.Metadata(ctx)
	if err != nil {
		t.Fatalf("metadata: %v", err)
	}
	if md.TimePartitioning == nil {
		t.Fatal("expected TimePartitioning, got nil")
	}
	if got := md.TimePartitioning.Type; got != bigquery.DayPartitioningType {
		t.Errorf("type = %v, want DAY", got)
	}
	if got := md.TimePartitioning.Field; got != "order_date" {
		t.Errorf("field = %v, want order_date", got)
	}
}

func TestClusteringMetadataRoundTrip(t *testing.T) {
	ctx := context.Background()
	client, _ := newClient(t)
	if err := client.Dataset("ds").Create(ctx, &bigquery.DatasetMetadata{}); err != nil {
		t.Fatalf("dataset: %v", err)
	}
	table := client.Dataset("ds").Table("sessions")
	if err := table.Create(ctx, &bigquery.TableMetadata{
		Schema: bigquery.Schema{
			{Name: "user_id", Type: bigquery.StringFieldType},
			{Name: "session_id", Type: bigquery.StringFieldType},
			{Name: "started_at", Type: bigquery.TimestampFieldType},
		},
		Clustering: &bigquery.Clustering{Fields: []string{"user_id", "session_id"}},
	}); err != nil {
		t.Fatalf("create: %v", err)
	}
	md, err := table.Metadata(ctx)
	if err != nil {
		t.Fatalf("metadata: %v", err)
	}
	if md.Clustering == nil {
		t.Fatal("expected Clustering, got nil")
	}
	want := []string{"user_id", "session_id"}
	if len(md.Clustering.Fields) != len(want) {
		t.Fatalf("fields = %v, want %v", md.Clustering.Fields, want)
	}
	for i, f := range md.Clustering.Fields {
		if f != want[i] {
			t.Errorf("fields[%d] = %v, want %v", i, f, want[i])
		}
	}
}

func TestIsPartitioningColumnViaInformationSchema(t *testing.T) {
	ctx := context.Background()
	client, project := newClient(t)
	if err := client.Dataset("ds").Create(ctx, &bigquery.DatasetMetadata{}); err != nil {
		t.Fatalf("dataset: %v", err)
	}
	table := client.Dataset("ds").Table("orders")
	if err := table.Create(ctx, &bigquery.TableMetadata{
		Schema: bigquery.Schema{
			{Name: "id", Type: bigquery.IntegerFieldType},
			{Name: "order_date", Type: bigquery.DateFieldType},
		},
		TimePartitioning: &bigquery.TimePartitioning{
			Type:  bigquery.DayPartitioningType,
			Field: "order_date",
		},
	}); err != nil {
		t.Fatalf("create: %v", err)
	}
	rows, err := readAll(t, client.Query(
		"SELECT column_name, is_partitioning_column "+
			"FROM `"+project+".ds`.INFORMATION_SCHEMA.COLUMNS "+
			"WHERE table_name = 'orders' ORDER BY ordinal_position"))
	if err != nil {
		t.Fatalf("query: %v", err)
	}
	byName := map[string]string{}
	for _, r := range rows {
		byName[r[0].(string)] = r[1].(string)
	}
	if byName["id"] != "NO" || byName["order_date"] != "YES" {
		t.Errorf("is_partitioning_column = %v, want {id:NO, order_date:YES}", byName)
	}
}

File: test/clients/java/src/test/java/local/bigquery/PartitioningTest.java
package local.bigquery;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.google.cloud.bigquery.BigQuery;
import com.google.cloud.bigquery.Clustering;
import com.google.cloud.bigquery.DatasetInfo;
import com.google.cloud.bigquery.Field;
import com.google.cloud.bigquery.FieldValueList;
import com.google.cloud.bigquery.InsertAllRequest;
import com.google.cloud.bigquery.Schema;
import com.google.cloud.bigquery.StandardSQLTypeName;
import com.google.cloud.bigquery.StandardTableDefinition;
import com.google.cloud.bigquery.Table;
import com.google.cloud.bigquery.TableId;
import com.google.cloud.bigquery.TableInfo;
import com.google.cloud.bigquery.TimePartitioning;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

/** Partitioning + clustering — mirrors partitioning_test.go. */
class PartitioningTest {
  @Test
  void ingestionTimePartitioning() throws Exception {
    BigQuery bq = Emu.client(Emu.uniqueProject());
    bq.create(DatasetInfo.newBuilder("ds").build());
    TableId tableId = TableId.of("ds", "events");
    bq.create(
        TableInfo.of(
            tableId,
            StandardTableDefinition.newBuilder()
                .setSchema(Schema.of(Field.of("kind", StandardSQLTypeName.STRING)))
                .setTimePartitioning(TimePartitioning.of(TimePartitioning.Type.DAY))
                .build()));
    bq.insertAll(
        InsertAllRequest.newBuilder(tableId)
            .addRow(Map.of("kind", "click"))
            .addRow(Map.of("kind", "view"))
            .build());

    List<FieldValueList> rows =
        Emu.rows(
            bq,
            "SELECT kind, _PARTITIONTIME IS NOT NULL AS has_ts FROM `ds.events` ORDER BY kind");
    for (FieldValueList r : rows) {
      assertTrue(r.get("has_ts").getBooleanValue(), () -> "_PARTITIONTIME missing");
    }
  }

  @Test
  void partitionDateFilterToday() throws Exception {
    BigQuery bq = Emu.client(Emu.uniqueProject());
    bq.create(DatasetInfo.newBuilder("ds").build());
    TableId tableId = TableId.of("ds", "events");
    bq.create(
        TableInfo.of(
            tableId,
            StandardTableDefinition.newBuilder()
                .setSchema(Schema.of(Field.of("kind", StandardSQLTypeName.STRING)))
                .setTimePartitioning(TimePartitioning.of(TimePartitioning.Type.DAY))
                .build()));
    bq.insertAll(
        InsertAllRequest.newBuilder(tableId)
            .addRow(Map.of("kind", "a"))
            .addRow(Map.of("kind", "b"))
            .build());

    List<FieldValueList> rows =
        Emu.rows(
            bq,
            "SELECT count(*) AS n FROM `ds.events` WHERE _PARTITIONDATE = CURRENT_DATE()");
    assertEquals(2L, rows.get(0).get("n").getLongValue());
  }

  @Test
  void columnPartitioningRoundTrip() {
    BigQuery bq = Emu.client(Emu.uniqueProject());
    bq.create(DatasetInfo.newBuilder("ds").build());
    TableId tableId = TableId.of("ds", "orders");
    bq.create(
        TableInfo.of(
            tableId,
            StandardTableDefinition.newBuilder()
                .setSchema(
                    Schema.of(
                        Field.of("id", StandardSQLTypeName.INT64),
                        Field.of("order_date", StandardSQLTypeName.DATE),
                        Field.of("amount", StandardSQLTypeName.FLOAT64)))
                .setTimePartitioning(
                    TimePartitioning.newBuilder(TimePartitioning.Type.DAY)
                        .setField("order_date")
                        .build())
                .build()));

    StandardTableDefinition def = bq.getTable(tableId).getDefinition();
    assertNotNull(def.getTimePartitioning());
    assertEquals(TimePartitioning.Type.DAY, def.getTimePartitioning().getType());
    assertEquals("order_date", def.getTimePartitioning().getField());
  }

  @Test
  void clusteringMetadataRoundTrip() {
    BigQuery bq = Emu.client(Emu.uniqueProject());
    bq.create(DatasetInfo.newBuilder("ds").build());
    TableId tableId = TableId.of("ds", "sessions");
    bq.create(
        TableInfo.of(
            tableId,
            StandardTableDefinition.newBuilder()
                .setSchema(
                    Schema.of(
                        Field.of("user_id", StandardSQLTypeName.STRING),
                        Field.of("session_id", StandardSQLTypeName.STRING),
                        Field.of("started_at", StandardSQLTypeName.TIMESTAMP)))
                .setClustering(
                    Clustering.newBuilder().setFields(List.of("user_id", "session_id")).build())
                .build()));

    StandardTableDefinition def = bq.getTable(tableId).getDefinition();
    assertNotNull(def.getClustering());
    assertEquals(List.of("user_id", "session_id"), def.getClustering().getFields());
  }

  @Test
  void isPartitioningColumnViaInformationSchema() throws Exception {
    String project = Emu.uniqueProject();
    BigQuery bq = Emu.client(project);
    bq.create(DatasetInfo.newBuilder("ds").build());
    bq.create(
        TableInfo.of(
            TableId.of("ds", "orders"),
            StandardTableDefinition.newBuilder()
                .setSchema(
                    Schema.of(
                        Field.of("id", StandardSQLTypeName.INT64),
                        Field.of("order_date", StandardSQLTypeName.DATE)))
                .setTimePartitioning(
                    TimePartitioning.newBuilder(TimePartitioning.Type.DAY)
                        .setField("order_date")
                        .build())
                .build()));

    List<FieldValueList> rows =
        Emu.rows(
            bq,
            "SELECT column_name, is_partitioning_column FROM `"
                + project
                + ".ds`.INFORMATION_SCHEMA.COLUMNS WHERE table_name = 'orders' ORDER BY ordinal_position");
    Map<String, String> byName = new HashMap<>();
    for (FieldValueList r : rows) {
      byName.put(r.get("column_name").getStringValue(), r.get("is_partitioning_column").getStringValue());
    }
    assertEquals("NO", byName.get("id"));
    assertEquals("YES", byName.get("order_date"));
  }
}
